examples/keras_recipes/distillation_recipe.py

```diff
@@ -0,0 +1,518 @@
+        self, optimizer, metrics, distillation_loss_fn, temperature=TEMPERATURE,
+    ):
+        super(Distiller, self).compile(optimizer=optimizer, metrics=metrics)
+        self.distillation_loss_fn = distillation_loss_fn
+        self.temperature = temperature
+
+    def train_step(self, data):
+        # Unpack data
+        x, _ = data
+
+        # Forward pass of teacher
+        teacher_predictions = self.teacher(x, training=False)
+
+        with tf.GradientTape() as tape:
+            # Forward pass of student
+            student_predictions = self.student(x, training=True)
+
+            # Compute loss
+            distillation_loss = self.distillation_loss_fn(
+                tf.nn.softmax(teacher_predictions / self.temperature, axis=1),
+                tf.nn.softmax(student_predictions / self.temperature, axis=1),
+            )
+
+        # Compute gradients
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(distillation_loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Report progress
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"distillation_loss": distillation_loss})
+        return results
+
+    def test_step(self, data):
+        # Unpack data
+        x, y = data
+
+        # Forward passes
+        teacher_predictions = self.teacher(x, training=False)
+        student_predictions = self.student(x, training=False)
+
+        # Calculate the loss
+        distillation_loss = self.distillation_loss_fn(
+            tf.nn.softmax(teacher_predictions / self.temperature, axis=1),
+            tf.nn.softmax(student_predictions / self.temperature, axis=1),
+        )
+
+        # Report progress
+        self.compiled_metrics.update_state(y, student_predictions)
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"distillation_loss": distillation_loss})
+        return results
+
+
+"""
+## Learning rate schedule
+
+A warmup cosine learning rate schedule is used in the paper. This schedule is also
+typical for many pre-training methods especially for computer vision.
+"""
+
+# Some code is referred from here:
+# https://www.kaggle.com/ashusma/training-rfcx-tensorflow-tpu-effnet-b2.
+
+
+class WarmUpCosine(keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(
+        self, learning_rate_base, total_steps, warmup_learning_rate, warmup_steps
+    ):
+        super(WarmUpCosine, self).__init__()
+
+        self.learning_rate_base = learning_rate_base
+        self.total_steps = total_steps
+        self.warmup_learning_rate = warmup_learning_rate
+        self.warmup_steps = warmup_steps
+        self.pi = tf.constant(np.pi)
+
+    def __call__(self, step):
+        if self.total_steps < self.warmup_steps:
+            raise ValueError("Total_steps must be larger or equal to warmup_steps.")
+        learning_rate = (
```

Comment:
That's a big paragraph-line to define the LR, consider breaking it up into a few separate lines / intermediate variables to make it more readable.